okhttp/src/main/kotlin/okhttp3/internal/platform/android/AndroidLog.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.platform.android
+
+import android.util.Log
+import java.util.concurrent.CopyOnWriteArraySet
+import java.util.logging.Handler
+import java.util.logging.Level
+import java.util.logging.LogRecord
+import java.util.logging.Logger
+import okhttp3.OkHttpClient
+import okhttp3.internal.SuppressSignatureCheck
+import okhttp3.internal.concurrent.TaskRunner
+import okhttp3.internal.http2.Http2
+
+private val LogRecord.androidLevel: Int
+  get() = when {
+    level.intValue() > Level.INFO.intValue() -> Log.WARN
+    level.intValue() == Level.INFO.intValue() -> Log.INFO
+    else -> Log.DEBUG
+  }
+
+@SuppressSignatureCheck
+object AndroidLog {
+  private const val MAX_LOG_LENGTH = 4000
+
+  internal fun androidLog(loggerName: String, logLevel: Int, message: String, t: Throwable?) {
+    val tag = knownLoggers[loggerName] ?: loggerName
+
+    if (Log.isLoggable(tag, logLevel)) {
+      var logMessage = message
+      if (t != null) logMessage = logMessage + '\n'.toString() + Log.getStackTraceString(t)
+
+      // Split by line, then ensure each line can fit into Log's maximum length.
+      var i = 0
+      val length = logMessage.length
+      while (i < length) {
+        var newline = logMessage.indexOf('\n', i)
+        newline = if (newline != -1) newline else length
+        do {
+          val end = minOf(newline, i + MAX_LOG_LENGTH)
+          Log.println(logLevel, tag, logMessage.substring(i, end))
+          i = end
+        } while (i < newline)
+        i++
+      }
+    }
+  }
+
+  // Keep references to loggers to prevent their configuration from being GC'd.
+  private val configuredLoggers = CopyOnWriteArraySet<Logger>()
+
+  private val knownLoggers = LinkedHashMap<String, String>().apply {
+    this[OkHttpClient::class.java.`package`.name] = "OkHttp"
+    this[OkHttpClient::class.java.name] = "okhttp.OkHttpClient"
+    this[Http2::class.java.name] = "okhttp.Http2"
+    this[TaskRunner::class.java.name] = "okhttp.TaskRunner"
+  }.toMap()
+
+  fun enable() {
+    val logger = Logger.getLogger(OkHttpClient::class.java.`package`.name)
+    if (configuredLoggers.add(logger)) {
+      // log everything and filter in androidLog method below
+      logger.level = Level.FINE
```

Comment:
@swankjesse If we are ok with having to call setprop before you run the app, then we can avoid log formatting.  But it just means you can't dynamically adjust after running.

Thoughts?

Reply:
I'll land it with a tweak so you need to restart the app after reloading, we can look at smarter options after that.  But it's really nice with just running

```
$ adb shell setprop log.tag.okhttp.Http2 DEBUG
```

Reply:
Is it gonna do string formatting on every frame?

Reply:
No. That's exactly what I flagged as I want to avoid.